corehq/apps/domain/models.py

```diff
@@ -622,6 +623,48 @@ def get_or_create_with_name(cls, name, is_active=False,
             new_domain.save(**get_safe_write_kwargs())
             return new_domain
 
+    @classmethod
+    def generate_name(cls, hr_name, max_length=25):
+        '''
+        Generate a URL-friendly name based on a given human-readable name.
+        Normalizes given name, then looks for conflicting domains, addressing
+        conflicts by adding "-1", "-2", etc. May return None if it fails to
+        generate a new, unique name. Throws exception if it can't figure out
+        a name, which shouldn't happen unless max_length is absurdly short.
+        '''
+
+        name = hr_name.strip().lower()
+        name = re.sub(r'[^0-9a-z]+', '-', name)
+
+        if Domain.get_by_name(name):
+            prefix = name
+            while len(prefix):
+                name = Domain._get_next_available_name(prefix, Domain.get_names_by_prefix(prefix + '-'))
+                if Domain.get_by_name(name):
+                    # should never happen
+                    raise Exception
```

Comment:
Open to using creating a custom exception if that's better practice. Given that it's only throw here and is immediately caught in forms.py, I took the lazier route.